flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/stream/StreamExecTableSourceScan.scala

```diff
@@ -173,7 +176,8 @@ class StreamExecTableSourceScan(
           val watermarkGenerator = new PeriodicWatermarkAssignerWrapper(rowtimeFieldIdx, p)
           ingestedTable.assignTimestampsAndWatermarks(watermarkGenerator)
         case p: PunctuatedWatermarkAssigner =>
-          val watermarkGenerator = new PunctuatedWatermarkAssignerWrapper(rowtimeFieldIdx, p)
+          val watermarkGenerator =
+            new PunctuatedWatermarkAssignerWrapper(rowtimeFieldIdx, p, producedDataType)
```

Comment:
Using `FlinkTypeFactory.toLogicalRowType(getRowType)` instead of `producedDataType`? So that we can avoid to do the pattern match in `PunctuatedWatermarkAssignerWrapper`.

Reply:
No, in this case, I want to keep user type information to an original Row. (Like a pojo field in Row)